Cross-Folder Registry

labkey-nicka opened the pull request:
#### Rationale
Component updates in support of cross-folder registry in LKB.

#### Related Pull Requests
* https://github.com/LabKey/labkey-ui-components/pull/725
* https://github.com/LabKey/biologics/pull/1139
* https://github.com/LabKey/platform/pull/3001
* https://github.com/LabKey/sampleManagement/pull/823
* https://github.com/LabKey/inventory/pull/372

#### Changes
* **AliasInput**
	* Refactor `AliasInput` "value" processing to match what is expected. This fixes several bugs with how this component displays and persists state.
	* Add unit tests.
* **Data Classes**
	* Support cross-folder domain editing for Data Classes.
* **Details**
	* Support `containerFilter` and `containerPath` for `Detail`, `DetailPanel`, `DetailEditRenderer`, and `EditableDetailPanel`.
* **Lineage**
	* Export `withLineage` and `InjectedLineage` for external use.
	* Refactor how `sampleStats` are processed.
* **Query APIs**
	* Improve typings for `insertRows`, `deleteRows`, and `selectRows` to reduce duplication with `@labkey/api` and provide passthrough of all configuration options.
	* Update `getContainerFilter` to process the `containerPath` provided to query configurations if provided.
	* Introduce `getContainerFilterForInsert`. Supplies the container filter to be used when fetching data intended for insert.
* **QuerySelect**
	* Ensure all queries made by `QuerySelect` specify the same set of columns.
	* Simplify and centralize `QuerySelectModel` initialization.
	* Support specifying `containerFilter` on `QuerySelect`.
* **SelectInput**
	* Introduce `resolveFormValue` for component users to override how the input processes the selected options into a form value.

labkey-nicka committed:
Export InjectedLineage, withLineage

labkey-nicka committed:
Update lineage sample stats

labkey-nicka committed:
Fix export of InjectedLineage

labkey-nicka committed:
ListResolver: prepare for cross-folder support

labkey-nicka committed:
getContainerFilterForInsert: support cf specific to insert
- Support containerFilter on QueryFormInputs and QuerySelect

labkey-nicka committed:
AliasInput: fix value resolution, form data binding
- add unit tests

labkey-nicka committed:
QuerySelect: ensure same columns in select, search requests

labkey-nicka committed:
Improve Query API wrapper typings

labkey-nicka committed:
SelectInput: add "resolveFormValue" prop
- allows for overridding the derivation of the form value from the selected options

labkey-nicka committed:
DetailPanel, et. al: Support containerPath, containerFilter

labkey-nicka committed:
SampleTypeLineageCounts: consolidate loading, error states

labkey-nicka committed:
ListResolver: support cross-folder list resolution

labkey-nicka committed:
DataClassDesigner: support cross-folder domain editing

cnathe commented on packages/components/src/internal/components/lineage/actions.ts:
```diff
@@ -284,54 +283,45 @@ export function loadSeedResult(seed: string, container?: string, options?: Linea
-            }
+        return {
+            name: {
+                url: AppURL.create('samples', name).toHref(),
```
minor: use the `SAMPLES_KEY` const

labkey-nicka replied:
Done.

cnathe commented on packages/components/src/internal/components/lineage/actions.ts:
```diff
@@ -284,54 +283,45 @@ export function loadSeedResult(seed: string, container?: string, options?: Linea
+            sampleCount: {
+                url:
+                    count > 0
+                        ? AppURL.create('rd', 'expdata', seedNode.id, 'samples').addParam('tab', name).toHref()
```
minor: use `SAMPLES_KEY` const

labkey-nicka replied:
This isn't actually the same thing. This is a new route that exists in LKB's registry to a new tab called "Samples" which displays all derived samples of a registry item.

cnathe replied:
Ah ok, got it. That makes sense.

cnathe commented on packages/components/src/internal/url/AppURLResolver.ts:
```diff
@@ -132,74 +132,91 @@ export class AssayRunResolver implements AppRouteResolver {
 /**
  * Resolves list routes dynamically
- * /q/lists/22/14/... -> /q/lists/listByName/14/...
+ * /q/lists/$CPS<container_path>$CPE/22/14/... -> /q/lists/listByName/14/...
```
interesting...I assume CPS = container path start and CPE is container path end, right?

cnathe replied:
and I assume this is needed because the paths have slashes in them? fun

labkey-nicka replied:
This probably isn't _needed_ but I couldn't create a regex that matched all the possibilities for what could be in a container path while still grouping the indices at the end.

labkey-nicka committed:
Review feedback

labkey-alan commented on packages/components/src/internal/components/lineage/actions.ts:
```diff
@@ -155,7 +165,7 @@ function applyItemMetadata(
 function applyLineageIOMetadata(
     item: LineageIOWithMetadata,
-    iconURLByLsid: { [lsid: string]: string },
+    iconURLByLsid: Record<string /* LSID */, string>,
```
We could consider something like:

`type LSID = string;`

Then make these `Record<LSID, string>`. It obviously won't make these things any more type safe than they already are from a compiler standpoint, but IDE Intellisense would at least tell you it's an LSID.

labkey-nicka committed:
Sample Status: Cross-folder support
- add "containerFilter" property to QueryLookup
- allow overriding QueryColumn lookup metadata

labkey-alan commented on packages/components/src/internal/url/AppURLResolver.ts:
```diff
@@ -132,74 +132,91 @@ export class AssayRunResolver implements AppRouteResolver {
+
+                    const _key = [_containerPath, _listId].join('|');
+                    return map.set(_key, _name);
+                }, Map<string, string>().asMutable())
```
You shouldn't need either the `asMutable()` or `asImmutable()` calls here since your reducer returns the result of `map.set`

labkey-nicka replied:
Yes, the difference is that with a mutable collection `map.set` will return the same instance as opposed to creating a new instance each time.

labkey-alan reviewed (approved):
Needs more references to `containerFilter`. I added a few small nits, otherwise looks good.

labkey-nicka committed:
lint

labkey-nicka committed:
Prepare release notes

labkey-nicka committed:
v2.129.0
